velox/functions/prestosql/window/NthValue.cpp

```diff
@@ -0,0 +1,189 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/exec/WindowFunction.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::window {
+
+namespace {
+
+template <typename T>
+class NthValueFunction : public exec::WindowFunction {
+ public:
+  explicit NthValueFunction(
+      const std::vector<exec::WindowFunctionArg>& args,
+      const TypePtr& resultType,
+      velox::memory::MemoryPool* pool)
+      : WindowFunction(resultType, pool), args_(args) {
+    VELOX_CHECK_EQ(args.size(), 2);
+    VELOX_CHECK_NULL(args[0].constantValue);
+    if (args[1].constantValue) {
+      isConstantOffset_ = true;
+      if (args[1].constantValue->value().isNull()) {
+        isConstantOffsetNull_ = true;
+        return;
+      }
+      constantOffset_ =
+          args[1].constantValue->value().template value<int64_t>();
+      VELOX_USER_CHECK_GE(constantOffset_, 1, "Offset must be at least 1");
+      return;
+    }
+    offsetsVector_ = BaseVector::create(BIGINT(), 0, pool);
+    offsetsFlatVector_ = offsetsVector_->template asFlatVector<int64_t>();
+  }
+
+  void resetPartition(const exec::WindowPartition* partition) {
+    partition_ = partition;
+    partitionOffset_ = 0;
+  }
+
+  template <bool constantOffset>
+  void getRowNumbersForOffset(
+      vector_size_t numRows,
+      const vector_size_t* frameStarts,
+      const vector_size_t* frameEnds) {
+    // The rowNumbers map for each output row, as per nth_value function
+    // semantics, the rowNumber (relative to the start of the partition) from
+    // which the input value should be copied.
+    // A rowNumber of -1 is for nullptr in the result.
+    rowNumbersVector_.resize(numRows);
+
+    if constexpr (constantOffset) {
+      if (isConstantOffsetNull_) {
+        for (int i = 0; i < numRows; i++) {
+          rowNumbersVector_[i] = -1;
+        }
+        return;
+      }
+      for (int i = 0; i < numRows; i++) {
+        auto frameStart = frameStarts[i];
+        auto frameEnd = frameEnds[i];
+        auto rowNumber = frameStart + constantOffset_ - 1;
+        rowNumbersVector_[i] = rowNumber <= frameEnd ? rowNumber : -1;
+      }
+      return;
+    }
+
+    offsetsVector_->resize(numRows);
+    partition_->extractColumn(
+        args_[1].argIndex, partitionOffset_, numRows, 0, offsetsVector_);
+    for (int i = 0; i < numRows; i++) {
+      if (offsetsFlatVector_->isNullAt(i)) {
+        rowNumbersVector_[i] = -1;
+      } else {
+        auto offset = offsetsFlatVector_->valueAt(i);
+        VELOX_USER_CHECK_GE(offset, 1, "Offset must be at least 1");
+        auto frameStart = frameStarts[i];
+        auto frameEnd = frameEnds[i];
+
+        auto rowNumber = frameStart + offset - 1;
+        rowNumbersVector_[i] = rowNumber <= frameEnd ? rowNumber : -1;
```

Comment:
nit: this logic repeats twice; would it make sense to extract it into a helper function

`rowNumbersVector_[i] = rowNumber(frameStarts[i], frameEnds[i], offset);`